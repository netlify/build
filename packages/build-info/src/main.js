import { getContext } from './context.js'
import { buildInfo } from './core.js'

export const getBuildInfo = async function (opts) {
  const context = await getContext(opts)
  return await buildInfo(context)
}
