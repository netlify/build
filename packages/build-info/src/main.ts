import { ContextOptions, getContext } from './context.js'
import { buildInfo } from './core.js'

export const getBuildInfo = async (opts: ContextOptions) => {
  const context = await getContext(opts)
  return await buildInfo(context)
}
