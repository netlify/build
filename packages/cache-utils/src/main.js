'use strict'

require('./utils/polyfills')

const del = require('del')

const { getCacheDir } = require('./dir')
const { moveCacheFile, hasFiles } = require('./fs')
const { list } = require('./list')
const { getManifestInfo, writeManifest, removeManifest, isExpired } = require('./manifest')
const { parsePath } = require('./path')

// Cache a file
const saveOne = async function (
  path,
  { move = DEFAULT_MOVE, ttl = DEFAULT_TTL, digests = [], cacheDir, cwd: cwdOpt, mode } = {},
) {
  const { srcPath, cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(srcPath))) {
    return false
  }

  const { manifestInfo, identical } = await getManifestInfo({ cachePath, move, ttl, digests })
  if (identical) {
    return true
  }

  await del(cachePath, { force: true })
  await moveCacheFile(srcPath, cachePath, move)
  await writeManifest(manifestInfo)

  return true
}

// Restore a cached file
const restoreOne = async function (path, { move = DEFAULT_MOVE, cacheDir, cwd: cwdOpt, mode } = {}) {
  const { srcPath, cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(cachePath))) {
    return false
  }

  if (await isExpired(cachePath)) {
    return false
  }

  await del(srcPath, { force: true })
  await moveCacheFile(cachePath, srcPath, move)

  return true
}

// Remove the cache of a file
const removeOne = async function (path, { cacheDir, cwd: cwdOpt, mode } = {}) {
  const { cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  if (!(await hasFiles(cachePath))) {
    return false
  }

  await del(cachePath, { force: true })
  await removeManifest(cachePath)

  return true
}

// Check if a file is cached
const hasOne = async function (path, { cacheDir, cwd: cwdOpt, mode } = {}) {
  const { cachePath } = await parsePath({ path, cacheDir, cwdOpt, mode })

  return (await hasFiles(cachePath)) && !(await isExpired(cachePath))
}

const DEFAULT_MOVE = false
const DEFAULT_TTL = undefined

// Allow each of the main functions to take either a single path or an array of
// paths as arguments
const allowMany = async function (func, paths, ...args) {
  if (!Array.isArray(paths)) {
    return func(paths, ...args)
  }

  const results = await Promise.all(paths.map((path) => func(path, ...args)))
  return results.some(Boolean)
}

const save = allowMany.bind(null, saveOne)
const restore = allowMany.bind(null, restoreOne)
const remove = allowMany.bind(null, removeOne)
const has = allowMany.bind(null, hasOne)

// Change `opts` default values
const bindOpts = function (opts) {
  return {
    save: (paths, optsA) => save(paths, { ...opts, ...optsA }),
    restore: (paths, optsA) => restore(paths, { ...opts, ...optsA }),
    remove: (paths, optsA) => remove(paths, { ...opts, ...optsA }),
    has: (paths, optsA) => has(paths, { ...opts, ...optsA }),
    list: (optsA) => list({ ...opts, ...optsA }),
    getCacheDir: (optsA) => getCacheDir({ ...opts, ...optsA }),
  }
}

module.exports = { save, restore, remove, has, list, getCacheDir, bindOpts }
