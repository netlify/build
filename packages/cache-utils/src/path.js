'use strict'

const { homedir } = require('os')
const { resolve, isAbsolute, join, sep } = require('path')

const { getCacheDir } = require('./dir')
const { safeGetCwd } = require('./utils/cwd')

// Find the paths of the file before/after caching
const parsePath = async function ({ path, cacheDir, cwdOpt, mode }) {
  const srcPath = await getSrcPath(path, cwdOpt)
  const cachePath = await getCachePath({ srcPath, cacheDir, cwdOpt, mode })
  return { srcPath, cachePath }
}

// Retrieve absolute path to the local file to cache/restore
const getSrcPath = async function (path, cwdOpt) {
  const cwd = await safeGetCwd(cwdOpt)
  const srcPath = resolvePath(path, cwd)
  checkSrcPath(srcPath, cwd)
  return srcPath
}

const resolvePath = function (path, cwd) {
  if (isAbsolute(path)) {
    return resolve(path)
  }

  if (cwd !== '') {
    return resolve(cwd, path)
  }

  throw new Error(`Current directory does not exist: ${cwd}`)
}

// Caching the whole repository creates many issues:
//  - It caches many directories that are not related to Gatsby but take lots of
//    space, such as node_modules
//  - It caches directories that are not meant to restored across builds. For
//    example .git (beside being big).
//  - It undoes any build operations inside the repository that might have
//    happened before this plugin starts restoring the cache, leading to
//    conflicts with other plugins, Netlify Build or the buildbot.
const checkSrcPath = function (srcPath, cwd) {
  if (cwd !== '' && isParentPath(srcPath, cwd)) {
    throw new Error(`Cannot cache ${srcPath} because it is the current directory (${cwd}) or a parent directory`)
  }
}

// Note: srcPath and cwd are already normalized and absolute
const isParentPath = function (srcPath, cwd) {
  return `${cwd}${sep}`.startsWith(`${srcPath}${sep}`)
}

const getCachePath = async function ({ srcPath, cacheDir, cwdOpt, mode }) {
  const cacheDirA = await getCacheDir({ cacheDir, mode })
  const { name, relPath } = await findBase(srcPath, cwdOpt)
  const cachePath = join(cacheDirA, name, relPath)
  return cachePath
}

// The cached path is the path relative to the base which can be either the
// current directory, the home directory or the root directory. Each is tried
// in order.
const findBase = async function (srcPath, cwdOpt) {
  const bases = await getBases(cwdOpt)
  const srcPathA = normalizeWindows(srcPath)
  return bases.map(({ name, base }) => parseBase(name, base, srcPathA)).find(Boolean)
}

// Windows drives are problematic:
//  - they cannot be used in `relPath` since directories cannot be called `C:`
//    for example.
//  - they make comparing between drives harder
// For the moment, we simply strip them. This does mean two files with the same
// paths but inside different drives would be cached together, which is not
// correct.
// This also means `cache.list()` always assumes the source files are on the
// current drive.
// TODO: fix it.
const normalizeWindows = function (srcPath) {
  return srcPath.replace(WINDOWS_DRIVE_REGEX, '\\')
}

const WINDOWS_DRIVE_REGEX = /^[a-zA-Z]:\\/

// This logic works when `base` and `path` are on different Windows drives
const parseBase = function (name, base, srcPath) {
  if (srcPath === base || !srcPath.startsWith(base)) {
    return
  }

  const relPath = srcPath.replace(base, '')
  return { name, relPath }
}

const getBases = async function (cwdOpt) {
  const cwdBase = await getCwdBase(cwdOpt)
  return [...cwdBase, { name: 'home', base: homedir() }, { name: 'root', base: sep }]
}

const getCwdBase = async function (cwdOpt) {
  const cwd = await safeGetCwd(cwdOpt)
  if (cwd === '') {
    return []
  }

  return [{ name: 'cwd', base: cwd }]
}

module.exports = { parsePath, getBases }
