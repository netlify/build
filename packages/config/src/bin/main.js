#!/usr/bin/env node
const process = require('process')

const { stableStringify } = require('fast-safe-stringify')

require('../utils/polyfills')

const { isUserError } = require('../error')
const resolveConfig = require('../main')

const { parseFlags } = require('./flags')

// CLI entry point
const runCli = async function() {
  try {
    const { stable, ...flags } = parseFlags()
    const result = await resolveConfig(flags)
    handleCliSuccess(result, stable)
  } catch (error) {
    handleCliError(error)
  }
}

// The result is printed as JSON on stdout on success (exit code 0)
const handleCliSuccess = function(result, stable) {
  const resultA = serializeApi(result)
  const stringifyFunc = stable ? stableStringify : JSON.stringify
  const resultJson = stringifyFunc(resultA, null, 2)
  console.log(resultJson)
  process.exitCode = 0
}

// `api` is not JSON-serializable, so we remove it
// We still indicate it as a boolean
const serializeApi = function({ api, ...result }) {
  if (api === undefined) {
    return result
  }

  return { ...result, hasApi: true }
}

const handleCliError = function(error) {
  // Errors caused by users do not show stack traces and have exit code 1
  if (isUserError(error)) {
    console.error(error.message)
    process.exitCode = 1
    return
  }

  // Internal errors / bugs have exit code 2
  console.error(error.stack)
  process.exitCode = 2
}

runCli()
