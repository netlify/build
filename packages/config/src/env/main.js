import omit from 'omit.js'

import { removeFalsy } from '../utils/remove_falsy.js'

import { getGitEnv } from './git.js'

// Retrieve this site's environment variable. Also take into account team-wide
// environment variables and addons.
// The buildbot already has the right environment variables. This is mostly
// meant so that local builds can mimic production builds
// TODO: add `netlify.toml` `build.environment`, after normalization
// TODO: add `CONTEXT` and others
export const getEnv = async function ({
  mode,
  config,
  siteInfo,
  accounts,
  addons,
  buildDir,
  branch,
  deployId,
  buildId,
  context,
}) {
  if (mode === 'buildbot') {
    return {}
  }

  const generalEnv = await getGeneralEnv({ siteInfo, buildDir, branch, deployId, buildId, context })
  const [accountEnv, addonsEnv, uiEnv, configFileEnv] = getUserEnv({ config, siteInfo, accounts, addons })

  // Sources of environment variables, in descending order of precedence.
  const sources = [
    { key: 'configFile', values: configFileEnv },
    { key: 'ui', values: uiEnv },
    { key: 'addons', values: addonsEnv },
    { key: 'account', values: accountEnv },
    { key: 'general', values: generalEnv },
  ]

  // A hash mapping names of environment variables to objects containing the following properties:
  // - sources: List of sources where the environment variable was found. The first element is the source that
  //   actually provided the variable (i.e. the one with the highest precedence).
  // - value: The value of the environment variable.
  const env = new Map()

  sources.forEach((source) => {
    Object.keys(source.values).forEach((key) => {
      if (env.has(key)) {
        const { sources: envSources, value } = env.get(key)

        env.set(key, {
          sources: [...envSources, source.key],
          value,
        })
      } else {
        env.set(key, {
          sources: [source.key],
          value: source.values[key],
        })
      }
    })
  })

  return Object.fromEntries(env)
}

// Environment variables not set by users, but meant to mimic the production
// environment.
const getGeneralEnv = async function ({
  siteInfo: { id, name, url, build_settings: { repo_url: REPOSITORY_URL } = {} },
  buildDir,
  branch,
  deployId,
  buildId,
  context,
}) {
  const gitEnv = await getGitEnv(buildDir, branch)
  return removeFalsy({
    SITE_ID: id,
    SITE_NAME: name,
    DEPLOY_ID: deployId,
    BUILD_ID: buildId,
    URL: url,
    REPOSITORY_URL,
    CONTEXT: context,
    NETLIFY_LOCAL: 'true',
    ...gitEnv,
    // Localization
    LANG: 'en_US.UTF-8',
    LANGUAGE: 'en_US:en',
    LC_ALL: 'en_US.UTF-8',
    // Disable telemetry of some tools
    GATSBY_TELEMETRY_DISABLED: '1',
    NEXT_TELEMETRY_DISABLED: '1',
  })
}

// Environment variables specified by the user
const getUserEnv = function ({ config, siteInfo, accounts, addons }) {
  const accountEnv = getAccountEnv({ siteInfo, accounts })
  const addonsEnv = getAddonsEnv(addons)
  const uiEnv = getUiEnv({ siteInfo })
  const configFileEnv = getConfigFileEnv({ config })
  return [accountEnv, addonsEnv, uiEnv, configFileEnv].map(cleanUserEnv)
}

// Account-wide environment variables
const getAccountEnv = function ({ siteInfo: { account_slug: accountSlug }, accounts }) {
  const { site_env: siteEnv = {} } = accounts.find(({ slug }) => slug === accountSlug) || {}
  return siteEnv
}

// Environment variables from addons
const getAddonsEnv = function (addons) {
  return Object.assign({}, ...addons.map(getAddonEnv))
}

const getAddonEnv = function ({ env }) {
  return env
}

// Site-specific environment variables set in the UI
const getUiEnv = function ({ siteInfo: { build_settings: { env = {} } = {} } }) {
  return env
}

// Site-specific environment variables set in netlify.toml
const getConfigFileEnv = function ({
  config: {
    build: { environment = {} },
  },
}) {
  return environment
}

// Some environment variables cannot be overridden by configuration
const cleanUserEnv = function (userEnv) {
  return omit.default(userEnv, READONLY_ENV)
}

const READONLY_ENV = [
  // Set in local builds
  'BRANCH',
  'CACHED_COMMIT_REF',
  'COMMIT_REF',
  'CONTEXT',
  'HEAD',
  'REPOSITORY_URL',
  'URL',

  // CI builds set NETLIFY=true while CLI and programmatic builds set
  // NETLIFY_LOCAL=true
  'NETLIFY',
  'NETLIFY_LOCAL',

  // Not set in local builds because there is no CI build/deploy, incoming hooks nor PR
  'BUILD_ID',
  'DEPLOY_ID',
  'DEPLOY_PRIME_URL',
  'DEPLOY_URL',
  'INCOMING_HOOK_BODY',
  'INCOMING_HOOK_TITLE',
  'INCOMING_HOOK_URL',
  'NETLIFY_BUILD_BASE',
  'NETLIFY_BUILD_LIFECYCLE_TRIAL',
  'NETLIFY_IMAGES_CDN_DOMAIN',
  'PULL_REQUEST',
  'REVIEW_ID',
]
