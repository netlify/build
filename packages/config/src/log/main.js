const { cleanupConfig } = require('./cleanup')
const { logObject, logSubHeader } = require('./logger')
const { cleanupConfigOpts } = require('./options')

// Log options in debug mode.
const logOpts = function(opts, { logs, debug, cachedConfig }) {
  // In production, print those in the first call to `@netlify/config`, not the
  // second one done inside `@netlify/build`
  if (!debug || cachedConfig !== undefined) {
    return
  }

  logSubHeader(logs, 'Initial build environment')
  logObject(logs, cleanupConfigOpts(opts))
}

// Log `defaultConfig` option in debug mode
const logDefaultConfig = function(defaultConfig, { logs, debug, baseRelDir }) {
  if (!debug || defaultConfig === undefined) {
    return
  }

  logSubHeader(logs, 'UI build settings')
  logObject(logs, cleanupConfig({ ...defaultConfig, baseRelDir }))
}

// Log `inlineConfig` option in debug mode
const logInlineConfig = function(initialConfig, { logs, debug }) {
  if (!debug) {
    return
  }

  logSubHeader(logs, 'CLI flags')
  logObject(logs, cleanupConfig(initialConfig))
}

// Log return value of `@netlify/config` in debug mode
const logResult = function({ configPath, buildDir, config, context, branch }, { logs, debug }) {
  if (!debug) {
    return
  }

  logSubHeader(logs, 'Resolved build environment')
  logObject(logs, { configPath, buildDir, context, branch })

  logSubHeader(logs, 'Resolved config')
  logObject(logs, cleanupConfig(config))
}

module.exports = { logOpts, logDefaultConfig, logInlineConfig, logResult }
