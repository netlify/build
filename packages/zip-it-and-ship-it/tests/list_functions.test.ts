import { join } from 'path'

import { describe, expect, test } from 'vitest'

import { listFunctions } from '../src/main.js'

import { FIXTURES_DIR, FIXTURES_ESM_DIR, normalizeFiles } from './helpers/main.js'

describe('listFunctions', () => {
  describe('v1', () => {
    test('Can list function main files with listFunctions()', async () => {
      const fixtureDir = `${FIXTURES_DIR}/list`
      const functions = await listFunctions(fixtureDir)

      expect(functions.map((file) => normalizeFiles(fixtureDir, file))).toMatchSnapshot()
    })

    test('Can list function main files from multiple source directories with listFunctions()', async () => {
      const fixtureDir = `${FIXTURES_DIR}/multiple-src-directories`
      const functions = await listFunctions([
        join(fixtureDir, '.netlify', 'internal-functions'),
        join(fixtureDir, 'netlify', 'functions'),
      ])

      expect(functions.map((file) => normalizeFiles(fixtureDir, file))).toMatchSnapshot()
    })

    test('listFunctions surfaces schedule config property', async () => {
      const functions = await listFunctions(join(FIXTURES_DIR, 'many-functions'), {
        config: {
          five: {
            schedule: '@daily',
          },
        },
      })
      const five = functions.find((func) => func.name === 'five')
      expect(five?.schedule).toBe('@daily')
    })

    test('listFunctions includes in-source config declarations', async () => {
      const fixtureDir = `${FIXTURES_DIR}/in-source-config/functions`
      const functions = await listFunctions(fixtureDir, {
        parseISC: true,
      })
      expect(functions.map((file) => normalizeFiles(fixtureDir, file))).toMatchSnapshot()
    })

    test('listFunctions includes json configured functions with configured properties', async () => {
      const dir = join(FIXTURES_DIR, 'json-config/.netlify/functions-internal/')
      const [func] = await listFunctions([dir], {
        configFileDirectories: [dir],
      })

      expect(func.displayName).toBe('A Display Name')
      expect(func.generator).toBe('@netlify/mock-plugin@1.0.0')
    })

    test('listFunctions does not include runtimeAPIVersion when parseISC is false', async () => {
      const dir = join(FIXTURES_DIR, 'list')
      const [func] = await listFunctions([dir], {
        parseISC: false,
      })

      expect(func.runtimeAPIVersion).toBeUndefined()
    })

    test('listFunctions includes runtimeAPIVersion when parseISC is true', async () => {
      const dir = join(FIXTURES_DIR, 'list')
      const [func] = await listFunctions([dir], {
        basePath: dir,
        parseISC: true,
      })

      expect(func.runtimeAPIVersion).toBe(1)
    })
  })

  describe('V2 API', () => {
    test('listFunctions does not include runtimeAPIVersion when parseISC is false', async () => {
      const dir = join(FIXTURES_ESM_DIR, 'v2-api')
      const [func] = await listFunctions([dir], {
        basePath: dir,
        parseISC: false,
      })

      expect(func.runtimeAPIVersion).toBeUndefined()
    })

    test('listFunctions includes runtimeAPIVersion when parseISC is true', async () => {
      const fixtureDir = join(FIXTURES_ESM_DIR, 'v2-api-esm-ts-aliases')
      const [func] = await listFunctions([fixtureDir], {
        basePath: fixtureDir,
        parseISC: true,
      })

      expect(normalizeFiles(fixtureDir, func)).toMatchSnapshot()
    })

    test('listFunctions includes routes when parseISC is true', async () => {
      const fixtureDir = join(FIXTURES_ESM_DIR, 'v2-api-with-path')
      const [func] = await listFunctions([fixtureDir], {
        basePath: fixtureDir,
        parseISC: true,
      })

      expect(normalizeFiles(fixtureDir, func)).toMatchSnapshot()
    })

    test('Accepts an object with mixed paths', async () => {
      const basePath = join(FIXTURES_ESM_DIR, 'v2-api-isolated')
      const functions = await listFunctions(
        {
          generated: {
            functions: [
              join(basePath, '.netlify/plugins/node_modules/extension-buildhooks/functions/extension-func1.mjs'),
              join(basePath, '.netlify/plugins/node_modules/extension-buildhooks/functions/extension-func2'),
            ],
          },
          user: {
            directories: [join(basePath, 'netlify/functions')],
          },
        },
        {
          basePath,
          parseISC: true,
        },
      )

      expect(functions.length).toBe(3)

      const userFunc1 = normalizeFiles(basePath, functions[0])
      expect(userFunc1.name).toBe('user-func1')
      expect(userFunc1.mainFile).toBe('netlify/functions/user-func1.mjs')

      const extensionFunc1 = normalizeFiles(basePath, functions[1])
      expect(extensionFunc1.name).toBe('extension-func1')
      expect(extensionFunc1.mainFile).toBe(
        '.netlify/plugins/node_modules/extension-buildhooks/functions/extension-func1.mjs',
      )

      const extensionFunc2 = normalizeFiles(basePath, functions[2])
      expect(extensionFunc2.name).toBe('extension-func2')
      expect(extensionFunc2.mainFile).toBe(
        '.netlify/plugins/node_modules/extension-buildhooks/functions/extension-func2/index.mjs',
      )
    })
  })
})
