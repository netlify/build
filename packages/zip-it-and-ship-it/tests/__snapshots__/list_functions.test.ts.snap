// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`listFunctions > V2 API > listFunctions includes routes when parseISC is true 1`] = `
{
  "displayName": undefined,
  "excludedRoutes": [
    {
      "expression": "^\\/products\\/sale(.*)\\/?$",
      "pattern": "/products/sale*",
    },
  ],
  "extension": ".mjs",
  "generator": undefined,
  "inputModuleFormat": "esm",
  "mainFile": "with-excluded-expression.mjs",
  "name": "with-excluded-expression",
  "routes": [
    {
      "expression": "^\\/products(?:\\/([^\\/]+?))\\/?$",
      "methods": [],
      "pattern": "/products/:id",
      "prefer_static": undefined,
    },
  ],
  "runtime": "js",
  "runtimeAPIVersion": 2,
  "schedule": undefined,
  "srcDir": "",
  "srcFile": undefined,
  "timeout": undefined,
}
`;

exports[`listFunctions > V2 API > listFunctions includes runtimeAPIVersion when parseISC is true 1`] = `
{
  "displayName": undefined,
  "excludedRoutes": [],
  "extension": ".ts",
  "generator": undefined,
  "inputModuleFormat": "esm",
  "mainFile": "function.ts",
  "name": "function",
  "routes": [],
  "runtime": "js",
  "runtimeAPIVersion": 2,
  "schedule": undefined,
  "srcDir": "",
  "srcFile": undefined,
  "timeout": undefined,
}
`;

exports[`listFunctions > v1 > Can list function main files from multiple source directories with listFunctions() 1`] = `
[
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": ".netlify/internal-functions/function.js",
    "name": "function",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": ".netlify/internal-functions",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": ".netlify/internal-functions/function_internal.js",
    "name": "function_internal",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": ".netlify/internal-functions",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "netlify/functions/function.js",
    "name": "function",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "netlify/functions",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "netlify/functions/function_user.js",
    "name": "function_user",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "netlify/functions",
    "srcFile": undefined,
    "timeout": undefined,
  },
]
`;

exports[`listFunctions > v1 > Can list function main files with listFunctions() 1`] = `
[
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".zip",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "test.zip",
    "name": "test",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "test.js",
    "name": "test",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".ts",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "five/index.ts",
    "name": "five",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "five",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "four.js/four.js.js",
    "name": "four",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "four.js",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "one/index.js",
    "name": "one",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "one",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "two/two.js",
    "name": "two",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "two",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": "",
    "generator": undefined,
    "inputModuleFormat": undefined,
    "mainFile": "test",
    "name": "test",
    "routes": undefined,
    "runtime": "go",
    "runtimeAPIVersion": undefined,
    "schedule": undefined,
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
]
`;

exports[`listFunctions > v1 > listFunctions includes in-source config declarations 1`] = `
[
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "cjs",
    "mainFile": "cron_cjs.js",
    "name": "cron_cjs",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "cjs",
    "mainFile": "cron_cjs_exports.js",
    "name": "cron_cjs_exports",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "cjs",
    "mainFile": "cron_cjs_renamed.js",
    "name": "cron_cjs_renamed",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm.js",
    "name": "cron_esm",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_additional_schedule_import.js",
    "name": "cron_esm_additional_schedule_import",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_no_direct_export.js",
    "name": "cron_esm_no_direct_export",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_no_direct_export_renamed.js",
    "name": "cron_esm_no_direct_export_renamed",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_no_direct_export_renamed_reassigned.js",
    "name": "cron_esm_no_direct_export_renamed_reassigned",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_renamed.js",
    "name": "cron_esm_renamed",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_schedule_multi_import.js",
    "name": "cron_esm_schedule_multi_import",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".js",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_esm_schedule_reassigned.js",
    "name": "cron_esm_schedule_reassigned",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".ts",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_ts.ts",
    "name": "cron_ts",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
  {
    "displayName": undefined,
    "excludedRoutes": undefined,
    "extension": ".ts",
    "generator": undefined,
    "inputModuleFormat": "esm",
    "mainFile": "cron_ts_renamed.ts",
    "name": "cron_ts_renamed",
    "routes": undefined,
    "runtime": "js",
    "runtimeAPIVersion": 1,
    "schedule": "@daily",
    "srcDir": "",
    "srcFile": undefined,
    "timeout": undefined,
  },
]
`;
